### ISSUE TITLE
Relaying errors when index row size exceeds maximum

### ISSUE THREAD

**Issue by saicitus** [citusdata/citus#1409]
Suppose I create a very large INDEX with some jsonb columns and there is a row(about to be inserted) which exceeds the maximum index size (2712), coordinator throws: 
**Error in the application:**
```
pg:could not modify any placements
```
**WARNING:**
```
postgresql-Wed.log-2017-05-10 21:11:01.022 UTC [17827][40/503] : [1-1] [app=[unknown]] WARNING:  index row size 5904 exceeds maximum 2712 for index "aggregates_pkey_102056"
postgresql-Wed.log-2017-05-10 21:11:01.022 UTC [17827][40/503] : [2-1] [app=[unknown]] HINT:  Values larger than 1/3 of a buffer page cannot be indexed.
postgresql-Wed.log-    Consider a function index of an MD5 hash of the value, or use full text indexing.
```

**Proposed solution:**
The error message could be improved by explicitly handling this case.

**Comment by anarazel** [comment]
On 2017-05-11 13:34:28 -0700, Sai wrote:
> Suppose I create a very large INDEX with some jsonb columns and there is a row(about to be inserted) which exceeds the maximum index size (2712), coordinator throws: 
> **Error in the application:**
> ```
> pg:could not modify any placements
> ```
> **WARNING:**
> ```
> postgresql-Wed.log-2017-05-10 21:11:01.022 UTC [17827][40/503] : [1-1] [app=[unknown]] WARNING:  index row size 5904 exceeds maximum 2712 for index "aggregates_pkey_102056"
> postgresql-Wed.log-2017-05-10 21:11:01.022 UTC [17827][40/503] : [2-1] [app=[unknown]] HINT:  Values larger than 1/3 of a buffer page cannot be indexed.
> postgresql-Wed.log-    Consider a function index of an MD5 hash of the value, or use full text indexing.
> ```
> 
> Proposed solution:
> The error message could be improved by explicitly handling this case.

How? And why's that worthwhile to do?

**Comment by lfittl** [comment]
@anarazel Question for clarification - I encountered a similar issue with a customer only seeing "could not modify any active placements" in their application/ORM's error message, without the actual error ("deadlock detected").

On the coordinator the output looked like this:

```
WARNING:  deadlock detected
DETAIL:  Process .. waits for ShareLock on transaction ..; blocked by process ...
        Process .. waits for ShareLock on transaction ..; blocked by process ...
HINT:  See server log for query details.
CONTEXT:  while executing command on ec2-1-1-1-1-compute-1.amazonaws.com:5432
        while inserting index tuple (...,..) in relation ...
ERROR:  could not modify any active placements
STATEMENT:  INSERT INTO ... ON CONFLICT ... DO UPDATE ... RETURNING (id)
```

What are the things we actually return to the client in such a case? Especially, are we returning both the WARNING and the ERROR, or just the ERROR?

**Comment by anarazel** [comment]
> @anarazel Question for clarification - I encountered a similar issue with a customer only seeing "could not modify any active placements" in their application/ORM's error message, without the actual error ("deadlock detected").

That's  #219 - which @sumedhpathak updated yesterday :)

> What are the things we actually return to the client in such a case? Especially, are we returning both the WARNING and the ERROR, or just the ERROR?

Both IIRC, but a lot of clients won't show WARNINGs.   The issue is that it's hard to discern failures due to the data from failures due to a dead node - which is important in the replication_factor > 1 cases, where we'd otherwise just mark a placement as invalid.

**Comment by lfittl** [comment]
I see. In that case I suggest closing this, as it seems to be that its a duplicate of https://github.com/citusdata/citus/issues/219

**Comment by ozgune** [comment]
@lfittl / @saicitus -- I'll refer to this issue in #219 and close this one.

Now that most of our deployments use streaming replication and we default to replication factor = 1, we could rethink how we emit these warning and error messages. Potentially, we could check if RF = 1, and error out early.
